src/syntax_analysis.rs

```diff
@@ -0,0 +1,192 @@
+use std::collections::HashMap;
+use syntect::parsing::syntax_definition::{
+    ContextReference, MatchOperation, MatchPattern, Pattern, SyntaxDefinition,
+};
+use syntect::parsing::{Scope, SyntaxSet, SyntaxSetBuilder};
+
+/// Used to look up (by name) what dependencies a given [SyntaxDefinition] has
+type SyntaxToDependencies = HashMap<String, Vec<Dependency>>;
+
+/// Used to look up what [SyntaxDefinition] that corresponds to a given [Dependency]
+type DependencyToSyntax<'a> = HashMap<Dependency, &'a SyntaxDefinition>;
+
+/// Represents a dependency on an external `.sublime-syntax` file.
+#[derive(Debug, Eq, PartialEq, Clone, Hash)]
+enum Dependency {
+    /// By name. Example YAML: `include: C.sublime-syntax`
+    ByName(String),
+
+    /// By scope. Example YAML: `embed: scope:source.c`
+    ByScope(Scope),
+}
+
+/// Change to true to make syntax dependency analysis print more details of what it is seeing
+const VERBOSE: bool = false;
+
+/// Generates independent [SyntaxSet]s after analyzing dependencies between syntaxes
+/// in a [SyntaxSetBuilder], and then prints the reults.
+pub(crate) fn print_syntax_dependencies(syntax_set_builder: &SyntaxSetBuilder) {
+    println!("Constructing independent SyntaxSets...");
+    let independent_syntax_sets = build_independent_syntax_sets(syntax_set_builder);
+
+    println!("Independent SyntaxSets:");
+    for syntax_set in independent_syntax_sets {
+        let names = syntax_set
+            .syntaxes()
+            .iter()
+            .map(|syntax| &syntax.name)
+            .collect::<Vec<&String>>();
+        println!("{:?}", names);
+    }
+}
+
+/// Analyzes dependencies between syntaxes in a [SyntaxSetBuilder].
+/// From that, it builds independent [SyntaxSet]s.
+fn build_independent_syntax_sets(syntax_set_builder: &SyntaxSetBuilder) -> Vec<SyntaxSet> {
```

Comment:
We definitely don't need to change this, as we are not optimizing for speed here. But the optimizer in me always gets a bit nervous when there is a superfluous container when the method could also return an iterator. Mostly for the curious:
```diff
--- a/src/syntax_analysis.rs
+++ b/src/syntax_analysis.rs
@@ -42,7 +42,7 @@ pub(crate) fn print_syntax_dependencies(syntax_set_builder: &SyntaxSetBuilder) {
 
 /// Analyzes dependencies between syntaxes in a [SyntaxSetBuilder].
 /// From that, it builds independent [SyntaxSet]s.
-fn build_independent_syntax_sets(syntax_set_builder: &SyntaxSetBuilder) -> Vec<SyntaxSet> {
+fn build_independent_syntax_sets<'a>(syntax_set_builder: &'a SyntaxSetBuilder) -> impl Iterator<Item=SyntaxSet> + 'a {
     let syntaxes = syntax_set_builder.syntaxes();
 
     // Build data structures we need for dependency resolution
@@ -51,7 +51,7 @@ fn build_independent_syntax_sets(syntax_set_builder: &SyntaxSetBuilder) -> Vec<S
     // Create one independent SyntaxSet from each (non-hidden) SyntaxDefinition
     syntaxes
         .iter()
-        .filter_map(|syntax| {
+        .filter_map(move |syntax| {
             if syntax.hidden {
                 return None;
             }
@@ -60,7 +60,6 @@ fn build_independent_syntax_sets(syntax_set_builder: &SyntaxSetBuilder) -> Vec<S
             builder.add_with_dependencies(syntax, &syntax_to_dependencies, &dependency_to_syntax);
             Some(builder.build())
         })
-        .collect()
 }
 
 /// In order to analyze dependencies, we need two key pieces of data.
```

Reply:
Cool! I would never have come up with that code myself. I don't quite understand how it can work actually. Seems to me as if something should be consumed and prevent later use when doing it like that. I'll have to study this in depth at a later point :)

But, the code works! And it results in nicer code, especially after `cargo fmt`, so I'll change to this instead :+1:

Reply:
FYI, I had to adjust the code slightly due to this lint warning:
```
warning: explicit lifetimes given in parameter types where they could be elided (or replaced with `'_` if needed by type declaration)
  --> src/syntax_dependencies.rs:43:1
   |
43 | / fn build_independent_syntax_sets<'a>(
44 | |     syntax_set_builder: &'a SyntaxSetBuilder,
45 | | ) -> impl Iterator<Item = SyntaxSet> + 'a {
   | |_________________________________________^
   |
   = note: `#[warn(clippy::needless_lifetimes)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes
```

Reply:
> Seems to me as if something should be consumed and prevent later use when doing it like that.

Not exactly sure what you mean, but yes: when using a `move |…| { … }` closure, something is consumed: the closure takes ownership of every captured variable. Which also means that they can not be used anywhere else after moving into the closure.